File: rp/validation/standard_validator.py
from validation.base_validator import BaseValidator
from dictionary.standard_dictionary import StandardDictionary
from typing import List
import re

_WORD_RE = re.compile(r"[^A-Za-z\-]")

class StandardValidator(BaseValidator):
    def __init__(self, dictionary: StandardDictionary):
        self.dictionary = dictionary

    def validate_chunk(self, chunk1: str, chunk2: str) -> bool:
        if chunk1[-1] == ' ' or chunk2[0] == ' ':
            return True

        words1 = chunk1.split()
        words2 = chunk2.split()

        if len(words1) < 2 or len(words2) < 2:
            return True

        combined = (words1[-1] + words2[0]).lower()
        return combined in self.dictionary.data

    def validate_text(self, text: str, required_chunks: List[str]) -> bool:
        for chunk in required_chunks:
            if chunk not in text:
                return False

        for raw in text.split():
            word = _WORD_RE.sub('', raw).lower()
            if word and word not in self.dictionary.data:
                return False

        return True

File: rp/text/validation/standard_validator.cpp
#include "standard_validator.hpp"
#include "../text_util.hpp"
#include <algorithm>
#include <cctype>

static std::string to_lower(std::string s) {
    std::transform(s.begin(), s.end(), s.begin(), [](unsigned char c) { return std::tolower(c); });
    return s;
}

bool StandardValidator::validate_chunk(const std::string& chunk1, const std::string& chunk2) const {
    if ((!chunk1.empty() && chunk1.back() == ' ') || (!chunk2.empty() && chunk2.front() == ' ')) {
        return true;
    }

    std::vector<std::string> words1 = split_words(chunk1);
    std::vector<std::string> words2 = split_words(chunk2);
    if (words1.size() < 2 || words2.size() < 2) return true;

    std::string combined = to_lower(words1.back() + words2.front());
    return dictionary.data.count(combined) > 0;
}

bool StandardValidator::validate_text(const std::string& text, const std::vector<std::string>& required_chunks) const {
    for (const std::string& chunk : required_chunks) {
        if (text.find(chunk) == std::string::npos) return false;
    }

    for (const std::string& raw : split_words(text)) {
        std::string word = clean_word(raw);
        if (!word.empty() && dictionary.data.count(word) == 0) return false;
    }

    return true;
}
